cpp/src/interpolate/cubic_spline.cu

```diff
@@ -0,0 +1,404 @@
+                  cudf::column_view const& prefixes,
+                  cudf::mutable_column_view const& D,
+                  cudf::mutable_column_view const& Dlu,
+                  cudf::mutable_column_view const& u,
+                  cudf::mutable_column_view const& h,
+                  cudf::mutable_column_view const& i,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      const T* T_ = t.data<T>();
+      const T* Y_ = y.data<T>();
+      const int32_t* PREFIXES = prefixes.data<int32_t>();
+      T* D_ = D.data<T>();
+      T* Dlu_ = Dlu.data<T>();
+      T* U_ = u.data<T>();
+      T* H_ = h.data<T>();
+      T* I_ = i.data<T>();
+      thrust::for_each(rmm::exec_policy(stream)->on(stream),
+        thrust::make_counting_iterator<int>(1),
+        thrust::make_counting_iterator<int>(prefixes.size()),
+        [T_, Y_, PREFIXES, D_, Dlu_, U_, H_, I_] __device__
+        (int index) {
+          int n = PREFIXES[index] - PREFIXES[index-1];
+          int h = PREFIXES[index-1];
+          int ci = 0;
+          for(ci = 0 ; ci < n-1 ; ++ci) {
+            H_[h + ci] = T_[h+ci+1] - T_[h+ci];
+            I_[h + ci] = (Y_[h+ci+1] - Y_[h+ci]) / H_[h + ci];
+          }
+          for(ci = 0 ; ci < n-2 ; ++ci) {
+            D_[h+ci+1] = (H_[h+ci+1]+H_[h+(n-2)-ci]) * 2;
+            U_[h+ci+1] = (I_[h+ci+1] - I_[h+(n-2)-ci]) * 6;
+          }
+          for(ci = 0 ; ci < n-3 ; ++ci) {
+            Dlu_[h+ci+1] = I_[h+ci+1];
+          }
+      });
+  }
+  template<typename T>
+  std::enable_if_t<not std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& y,
+                  cudf::column_view const& prefixes,
+                  cudf::mutable_column_view const& D,
+                  cudf::mutable_column_view const& Dlu,
+                  cudf::mutable_column_view const& u,
+                  cudf::mutable_column_view const& h,
+                  cudf::mutable_column_view const& i,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      CUDF_FAIL("Non-floating point operation is not supported.");
+  }
+};
+
+} // anonymous namespace
+
+namespace cuspatial
+{
+
+std::unique_ptr<cudf::column> cubicspline_interpolate(
+    cudf::column_view const& query_points,
+    cudf::column_view const& curve_ids,
+    cudf::column_view const& prefixes,
+    cudf::column_view const& source_points,
+    cudf::table_view const& coefficients
+)
+{
+    cudaStream_t stream=0;
+    rmm::mr::device_memory_resource* mr=rmm::mr::get_default_resource();
+    auto result = make_numeric_column(query_points.type(), query_points.size(), cudf::mask_state::UNALLOCATED, stream, mr);
+    cudf::mutable_column_view search_result = result->mutable_view();
+    
+    parallel_search search;
+    search.operator()<float>(query_points, curve_ids, prefixes, source_points, search_result, mr, stream);
```

Comment:
Make search and element wise functor and move the thrust::transform here and pass search to it.  Same for interpolate below.